File: assignments/assignment-1-solution/src/CrazyEights/Domain/PlayableCardsSelector.cs
using CrazyEights.Cards;

namespace CrazyEights.Domain;

public static class PlayableCardsSelector
{
    public static IReadOnlyList<PlayableCard> Get(
        IReadOnlyList<ICard> cards,
        SuitType currentSuit,
        RankType currentRank)
    {
        var playableCards = new List<PlayableCard>();
        foreach (var card in cards)
        {
            var outcome = CanPlayCard(card, currentSuit, currentRank);
            if (!outcome.CanPlay)
                continue;

            var playableCard = new PlayableCard(card, outcome.Reason);
            playableCards.Add(playableCard);
        }

        return playableCards.AsReadOnly();
    }

    public static IsPlayable CanPlayCard(ICard card, SuitType currentSuitType, RankType currentRankType)
    {
        if (Rank.IsWildcardRank(card.Rank))
            return new IsPlayable(true, "Wildcard!");

        if (currentRankType == card.Rank)
            return new IsPlayable(true, "Matches Rank");

        if (currentSuitType == card.Suit)
            return new IsPlayable(true, "Matches Suit");

        return new IsPlayable(false, "Not Playable");
    }
}

File: assignments/assignment-1-solution/src-java/src/main/java/edu/kennesaw/crazy8s/domain/PlayableCardsSelector.java
package edu.kennesaw.crazy8s.domain;

import edu.kennesaw.crazy8s.cards.Card;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Determines which cards can be played given the current suit and rank.
 */
public final class PlayableCardsSelector {
    private PlayableCardsSelector() {
    }

    public static List<PlayableCard> get(List<Card> cards, SuitType currentSuit, RankType currentRank) {
        List<PlayableCard> playableCards = new ArrayList<>();
        for (Card card : cards) {
            IsPlayable outcome = canPlayCard(card, currentSuit, currentRank);
            if (!outcome.canPlay()) {
                continue;
            }

            PlayableCard playableCard = new PlayableCard(card, outcome.getReason());
            playableCards.add(playableCard);
        }

        return Collections.unmodifiableList(playableCards);
    }

    public static IsPlayable canPlayCard(Card card, SuitType currentSuitType, RankType currentRankType) {
        if (Rank.isWildcardRank(card.getRank())) {
            return new IsPlayable(true, "Wildcard!");
        }

        if (currentRankType == card.getRank()) {
            return new IsPlayable(true, "Matches Rank");
        }

        if (currentSuitType == card.getSuit()) {
            return new IsPlayable(true, "Matches Suit");
        }

        return new IsPlayable(false, "Not Playable");
    }
}
